src/async_wrap.cc

```diff
@@ -310,23 +219,18 @@ static void SetupHooks(const FunctionCallbackInfo<Value>& args) {
   SET_HOOK_FN(destroy);
   SET_HOOK_FN(promise_resolve);
 #undef SET_HOOK_FN
+}
 
-  {
-    Local<FunctionTemplate> ctor =
-        FunctionTemplate::New(env->isolate());
-    ctor->SetClassName(FIXED_ONE_BYTE_STRING(env->isolate(), "PromiseWrap"));
-    Local<ObjectTemplate> promise_wrap_template = ctor->InstanceTemplate();
-    promise_wrap_template->SetInternalFieldCount(
-        PromiseWrap::kInternalFieldCount);
-    promise_wrap_template->SetAccessor(
-        FIXED_ONE_BYTE_STRING(env->isolate(), "isChainedPromise"),
-        PromiseWrap::getIsChainedPromise);
-    env->set_promise_wrap_template(promise_wrap_template);
+static void EnablePromiseHook(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  if (!args[0]->IsFunction()) {
+    return node::THROW_ERR_INVALID_ARG_TYPE(
+        env, "handler must be a function");
   }
-}
 
+  env->set_promise_hook_handler(args[0].As<Function>());
```

Comment:
Does it make sense to clean up `promise_hook_handler` in `DisablePromiseHook` as counterpart of this initialization?

Reply:
Yep, I'll get to that. Keep in mind this is still very much a work-in-progress PR, so expect lots of missing details as I work out if the general idea is reasonable. 😅